File: examples/kotlin/examples/ClientSchedules.kt
package io.ekodb.client.examples

import io.ekodb.client.EkoDBClient
import io.ekodb.client.functions.FunctionStageConfig
import io.ekodb.client.functions.UserFunction
import io.github.cdimascio.dotenv.dotenv
import kotlinx.coroutines.runBlocking
import kotlinx.serialization.json.buildJsonObject
import kotlinx.serialization.json.boolean
import kotlinx.serialization.json.jsonArray
import kotlinx.serialization.json.jsonObject
import kotlinx.serialization.json.jsonPrimitive
import kotlinx.serialization.json.put

/**
 * Schedule management example.
 *
 * Exercises create -> list -> get -> update -> trigger -> pause -> resume -> delete.
 * A temporary no-op function makes the trigger step self-contained.
 */
fun main() = runBlocking {
    val dotenv = dotenv()
    val baseUrl = dotenv["API_BASE_URL"] ?: "http://localhost:8080"
    val apiKey = dotenv["API_BASE_KEY"] ?: "a-test-api-key-from-ekodb"

    val client = EkoDBClient.builder()
        .baseUrl(baseUrl)
        .apiKey(apiKey)
        .build()

    println("=== ekoDB Kotlin Client - Schedules Example ===\n")

    var functionId: String? = null
    var schedId: String? = null
    var primaryError: Throwable? = null
    try {
        val functionLabel = "schedule_noop_kotlin_${ProcessHandle.current().pid()}_${System.currentTimeMillis()}"
        functionId = client.saveFunction(UserFunction(
            label = functionLabel,
            name = "Schedule example no-op",
            parameters = emptyMap(),
            functions = listOf(
                FunctionStageConfig.Return(
                    fields = mapOf("triggered" to kotlinx.serialization.json.JsonPrimitive(true)),
                    status_code = 200
                )
            )
        ))

        // 1. Create a schedule
        println("--- Creating schedule ---")
        val schedule = client.createSchedule(buildJsonObject {
            put("name", "Hourly Health Check")
            put("function_label", functionLabel)
            put("cron_expression", "0 0 * * * *")
            put("description", "Ping all services every hour")
            put("enabled", true)
        })
        schedId = requireNotNull(schedule["id"]) { "Created schedule omitted id" }.jsonPrimitive.content
        check(schedule["name"]?.jsonPrimitive?.content == "Hourly Health Check") {
            "Unexpected created schedule name: ${schedule["name"]}"
        }
        check(schedule["cron_expression"]?.jsonPrimitive?.content == "0 0 * * * *") {
            "Unexpected created schedule cron: ${schedule["cron_expression"]}"
        }
        check(schedule["enabled"]?.jsonPrimitive?.boolean == true) {
            "Created schedule was not enabled: ${schedule["enabled"]}"
        }
        println("Created schedule: $schedId — ${schedule["name"]}")

        // 2. List all schedules
        println("\n--- Listing schedules ---")
        val list = client.listSchedules()
        println("Schedules: $list")
        val listedIds = requireNotNull(list["schedules"]) {
            "Schedule list response omitted schedules"
        }.jsonArray.map { item ->
            requireNotNull(item.jsonObject["id"]) { "Listed schedule omitted id" }.jsonPrimitive.content
        }
        check(schedId in listedIds) { "Created schedule $schedId was absent from listSchedules" }

        // 3. Get schedule by ID
        println("\n--- Getting schedule ---")
        val fetched = client.getSchedule(schedId)
        println("Fetched: ${fetched["name"]} (cron: ${fetched["cron_expression"]})")
        check(fetched["id"]?.jsonPrimitive?.content == schedId) {
            "Fetched schedule had unexpected id: ${fetched["id"]}"
        }
        check(fetched["cron_expression"]?.jsonPrimitive?.content == "0 0 * * * *") {
            "Fetched schedule had unexpected cron: ${fetched["cron_expression"]}"
        }

        // 4. Update schedule
        println("\n--- Updating schedule ---")
        val updated = client.updateSchedule(schedId, buildJsonObject {
            put("cron_expression", "0 */30 * * * *")
            put("description", "Ping all services every 30 minutes")
        })
        println("Updated cron: ${updated["cron_expression"]}")
        check(updated["cron_expression"]?.jsonPrimitive?.content == "0 */30 * * * *") {
            "Schedule update did not persist cron: ${updated["cron_expression"]}"
        }

        // 5. Trigger immediately
        println("\n--- Triggering schedule ---")
        val triggered = client.triggerSchedule(schedId)
        println("Trigger response: $triggered")
        check(triggered["schedule_id"]?.jsonPrimitive?.content == schedId) {
            "Trigger response had unexpected schedule id: ${triggered["schedule_id"]}"
        }

        // 6. Pause schedule
        println("\n--- Pausing schedule ---")
        val paused = client.pauseSchedule(schedId)
        println("Enabled after pause: ${paused["enabled"]}")
        check(paused["enabled"]?.jsonPrimitive?.boolean == false) {
            "Pause left schedule enabled: ${paused["enabled"]}"
        }

        // 7. Resume schedule
        println("\n--- Resuming schedule ---")
        val resumed = client.resumeSchedule(schedId)
        println("Enabled after resume: ${resumed["enabled"]}")
        check(resumed["enabled"]?.jsonPrimitive?.boolean == true) {
            "Resume left schedule disabled: ${resumed["enabled"]}"
        }

        // 8. Delete schedule
        println("\n--- Deleting schedule ---")
    } catch (error: Throwable) {
        primaryError = error
        throw error
    } finally {
        val cleanupErrors = mutableListOf<Throwable>()
        schedId?.let { id ->
            try {
                client.deleteSchedule(id)
                println("Schedule deleted successfully")
            } catch (error: Throwable) {
                cleanupErrors.add(error)
            }
        }
        functionId?.let { id ->
            try {
                client.deleteFunction(id)
            } catch (error: Throwable) {
                cleanupErrors.add(error)
            }
        }
        try {
            client.close()
        } catch (error: Throwable) {
            cleanupErrors.add(error)
        }

        val failure = primaryError
        if (failure != null) {
            cleanupErrors.forEach(failure::addSuppressed)
        } else if (cleanupErrors.isNotEmpty()) {
            val cleanupError = cleanupErrors.first()
            cleanupErrors.drop(1).forEach(cleanupError::addSuppressed)
            throw cleanupError
        } else {
            println("\n=== Example Complete ===")
        }
    }
}

File: examples/rust/examples/client_schedules.rs
//! ekoDB Rust Client - Schedule Management Example
//!
//! Demonstrates the full schedule lifecycle: create, list, get, update,
//! trigger, pause, resume, and delete. It creates a temporary no-op function
//! so the immediate trigger exercises a complete, standalone lifecycle.
//!
//! Run with: `cargo run --example client_schedules`

use ekodb_client::{Client, Function, UserFunction};
use serde_json::json;
use std::collections::HashMap;
use std::error::Error;

#[tokio::main]
async fn main() -> Result<(), Box<dyn Error>> {
    dotenv::dotenv().ok();

    println!("=== ekoDB Schedule Management Example (Rust) ===\n");

    let client = Client::builder()
        .base_url(
            std::env::var("API_BASE_URL").unwrap_or_else(|_| "http://localhost:8080".to_string()),
        )
        .api_key(std::env::var("API_BASE_KEY")?)
        .build()?;

    let function_label = format!("schedule_noop_rust_{}", std::process::id());
    let function_id = client
        .save_function(
            UserFunction::new(&function_label, "Schedule example no-op").with_function(
                Function::Return {
                    fields: HashMap::from([("triggered".to_string(), json!(true))]),
                    status_code: Some(200),
                },
            ),
        )
        .await?;

    let mut schedule_id = None;
    let operations: Result<(), Box<dyn Error>> = async {
        // 1. Create a schedule
        println!("--- Creating schedule ---");
        let schedule = client
            .create_schedule(json!({
                "name": "nightly-cleanup",
                "description": "Remove expired sessions and temp files",
                "function_label": function_label,
                "cron_expression": "0 0 3 * * *",
                "timezone": "UTC",
                "enabled": true,
            }))
            .await?;
        schedule_id = Some(schedule["id"].as_str().unwrap_or_default().to_string());
        let schedule_id = schedule_id.as_deref().unwrap_or_default();
        println!(
            "Created schedule: {} (id: {})",
            schedule["name"], schedule_id
        );

        // 2. List all schedules
        println!("\n--- Listing schedules ---");
        let schedules = client.list_schedules().await?;
        println!("Schedules: {}", schedules);

        // 3. Get schedule by ID
        println!("\n--- Getting schedule ---");
        let fetched = client.get_schedule(&schedule_id).await?;
        println!("Fetched schedule: {}", fetched["name"]);

        // 4. Update schedule
        println!("\n--- Updating schedule ---");
        let updated = client
            .update_schedule(
                &schedule_id,
                json!({
                    "description": "Remove expired sessions, temp files, and orphaned uploads",
                    "cron_expression": "0 0 4 * * *",
                }),
            )
            .await?;
        println!("Updated cron: {}", updated["cron_expression"]);
        println!("Updated description: {}", updated["description"]);

        // 5. Trigger the schedule immediately
        println!("\n--- Triggering schedule ---");
        let triggered = client.trigger_schedule(&schedule_id).await?;
        println!("Trigger response: {}", triggered);

        // 6. Pause schedule
        println!("\n--- Pausing schedule ---");
        let paused = client.pause_schedule(&schedule_id).await?;
        println!("Schedule paused: {}", paused);

        // 7. Resume schedule
        println!("\n--- Resuming schedule ---");
        let resumed = client.resume_schedule(&schedule_id).await?;
        println!("Schedule resumed: {}", resumed);

        // 8. Delete schedule
        println!("\n--- Deleting schedule ---");
        Ok(())
    }
    .await;

    let schedule_cleanup = match schedule_id.as_deref() {
        Some(id) => client.delete_schedule(id).await,
        None => Ok(()),
    };
    let function_cleanup = client.delete_function(&function_id).await;
    operations?;
    schedule_cleanup?;
    println!("Schedule deleted successfully");
    function_cleanup?;

    println!("\n=== All schedule operations completed ===");
    Ok(())
}
